runtime/wasm/src/host_exports.rs

```diff
@@ -335,6 +343,53 @@ where
         )
     }
 
+    // Read the IPFS file `link`, split it into JSON objects, and invoke
+    // the exported function `callback` on each JSON object. The successful
+    // return value contains all entity operations that were produced by the
+    // callback invocations. Each invocation of `callback` happens in its own
+    // instance of a WASM module, which is identical to `module` when it was
+    // first started.
+    pub(crate) fn ipfs_map(
+        &self,
+        module: &WasmiModule<E, L, S, U>,
+        link: String,
+        callback: &str,
+        flags: Vec<String>,
+    ) -> Result<Vec<EntityOperation>, HostExportError<impl ExportError>> {
+        const JSON_FLAG: &str = "json";
+        if !flags.contains(&JSON_FLAG.to_string()) {
+            return Err(HostExportError(format!("Flags must contain 'json'")));
+        }
+
+        let valid_module = module.valid_module.clone();
+        let ctx = module.ctx.clone();
+        let callback = callback.to_owned();
+        // Create a base error message to avoid borrowing headaches
+        let errmsg = format!(
+            "ipfs_map: callback '{}' failed when processing file '{}'",
+            &*callback, &link
+        );
+
+        let operations = self.block_on(
+            self.link_resolver
+                .json_stream(&Link { link })
+                .and_then(|stream| {
+                    stream
+                        .and_then(move |v| {
```

Comment:
I'd use `fold` here instead of `and_then`. That way you don't have to do the `into_iter().flatten().collect()` dance at the end.

Essentially, something like this:
```
stream
    .fold(vec![], |mut operations, v| {
        ...
        let new_operations = module.handle_json_callback(&*callback, &v)?;
        v.extend(&new_operations);
        v
    })
```

Reply:
Maybe `iter().flatten().collect()` is more efficient even? I'm not sure. Just whenever I see a stream or array being turned into a single result value (e.g. a vector), I think of it as a reducing/folding operation.

Reply:
I think [concat](https://docs.rs/futures/0.1.17/futures/stream/trait.Stream.html#method.concat2) may work here.

Reply:
If I use `fold` here, i.e., turn the code to
```
        let operations = self.block_on(
            self.link_resolver
                .json_stream(&Link { link })
                .and_then(|stream| {
                    stream.fold(vec![], move |mut ops, v| {
                        let module = WasmiModule::from_valid_module_with_ctx(
                            valid_module.clone(),
                            ctx.clone(),
                        )?;
                        ops.extend(module.handle_json_callback(&*callback, &v)?);
                        Ok(ops)
                    })
                })
                .map_err(move |e| HostExportError(format!("{}: {}", errmsg, e.to_string()))),
        )?;
```

I get an error
```
error[E0283]: type annotations required: cannot resolve `graph::prelude::Error: std::convert::From<_>`
   --> runtime/wasm/src/host_exports.rs:377:28
    |
377 |                     stream.fold(vec![], move |mut ops, v| {
    |                            ^^^^

error: aborting due to previous error
```

and I am not quite sure what rustc is trying to tell me here.

Reply:
I wouldn't worry about it in this case — I bet the flatten approach that you've picked isn't going to be a problem. @leodasvacas What do you think?

Reply:
The error is misterious to me as well, and I also don't think the current approach is an issue.